librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+        }
+        if data.keys.is_empty() {
+            return Err(Error::InvalidData("Missing keys".to_owned()));
+        }
+
+        let mut keys = HashSet::new();
+        for k in data.keys.iter() {
+            keys.insert(
+                PublicKey::from_bs58(k).ok_or_else(|| Error::InvalidData(format!("key: {}", k)))?,
+            );
+        }
+
+        let mut certifiers = HashSet::new();
+        for c in data.certifiers.iter() {
+            certifiers.insert(
+                RadicleUri::from_str(c)
+                    .map_err(|_| Error::InvalidData(format!("certifier: {}", c)))?,
+            );
+        }
+
+        let mut signatures = HashMap::new();
+        if let Some(s) = &data.signatures {
+            for (k, sig) in s.iter() {
+                let key = PublicKey::from_bs58(k)
+                    .ok_or_else(|| Error::InvalidData(format!("signature key: {}", k)))?;
+                let signature = EntitySignature {
+                    by: match &sig.user {
+                        Some(uri) => Signatory::User(RadicleUri::from_str(&uri)?),
+                        None => Signatory::OwnedKey,
+                    },
+                    sig: Signature::from_bs58(&sig.sig).ok_or_else(|| {
+                        Error::InvalidData(format!("signature data: {}", &sig.sig))
+                    })?,
+                };
+                signatures.insert(key, signature);
+            }
+        }
+
+        let actual_hash = data.compute_hash()?;
+        if let Some(s) = &data.hash {
+            let claimed_hash = {
+                let bytes = bs58::decode(s.as_bytes())
+                    .with_alphabet(bs58::alphabet::BITCOIN)
+                    .into_vec()
+                    .map_err(|_| Error::InvalidBufferEncoding(s.to_owned()))?;
+                Multihash::from_bytes(bytes).map_err(|_| Error::InvalidHash(s.to_owned()))?
+            };
+            if claimed_hash != actual_hash {
+                return Err(Error::InvalidHash(s.to_owned()));
+            }
+        }
+
+        let parent_hash = match data.parent_hash {
+            Some(s) => {
+                let bytes = bs58::decode(s.as_bytes())
+                    .with_alphabet(bs58::alphabet::BITCOIN)
+                    .into_vec()
+                    .map_err(|_| Error::InvalidBufferEncoding(s.to_owned()))?;
+                let hash =
+                    Multihash::from_bytes(bytes).map_err(|_| Error::InvalidHash(s.to_owned()))?;
+                Some(hash)
+            },
+            None => None,
+        };
+
+        Ok(Self {
+            name: data.name.unwrap(),
+            revision: data.revision.unwrap().to_owned(),
+            hash: actual_hash,
+            parent_hash,
+            keys,
+            certifiers,
+            signatures,
+            info: data.info,
+        })
+    }
+
+    /// Turn the entity in to its raw data
+    /// (first step of serialization and reverse of `from_data`)
```

Comment:
This creates a hyperlink to the function in the rendered documentation:
```suggestion
    /// (first step of serialization and reverse of [`Entity::from_data`])
```